File: js/problem32.js
/**
 * @param {string} s
 * @return {number}
 */
var longestValidParentheses = function(s) {
    if (s.length <= 1) {
        return 0;
    }
    
    let stack = [-1], res = 0;
    
    for (let i = 0; i < s.length; i++) {
        if (s[i] === '(') {
            stack.push(i);
        } else {
            stack.pop();
            if (stack.length === 0) {
                stack.push(i);
            } else {
                let temp = (i + 1) - stack[stack.length - 1] - 1;
                if (temp > res) {
                    res = temp;
                }
            }
        }
    }
    
    return res;
    
};

File: leetcodeJava/src/main/java/Problem32.java
// 32. Longest Valid Parentheses

class Solution {
    public int longestValidParentheses(String s) {
        if (s.length() <= 1) {
            return 0;
        }
        
        int res = 0;
        Stack<Integer> stack = new Stack<Integer>();
        stack.push(-1);
        
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                stack.push(i);
            } else {
                int temp = stack.pop();
                if (!stack.empty()) {
                    int currentMaxLength = i - stack.peek();
                    res = (currentMaxLength > res) ? currentMaxLength : res;
                } else {
                    stack.push(i);
                }
            }
        }
        
        return res;
    }
}